File: 2395-LongestBinarySubsequenceLessThanOrEqualToK/2395-LongestBinarySubsequenceLessThanOrEqualToK.java
// Last updated: 6/16/2026, 8:51:14 AM
public class Solution {
    public int longestSubsequence(String s, int k) {
        long currentVal = 0;
        int ansLength = 0;
        long power = 1;

        for (int i = s.length() - 1; i >= 0; i--) {
            char ch = s.charAt(i);

            if (ch == '0') {
                ansLength++;
            } else {
                if (currentVal + power <= k) {
                    currentVal += power;
                    ansLength++;
                }
            }

            if (power <= k) {
                power *= 2;
            }
        }

        return ansLength;
    }
}

File: 2395-LongestBinarySubsequenceLessThanOrEqualToK/2395-LongestBinarySubsequenceLessThanOrEqualToK.py
# Last updated: 6/16/2026, 8:51:20 AM
class Solution:
    def longestSubsequence(self, s: str, k: int) -> int:
        current_val = 0
        ans_length = 0
        power = 1

        for i in range(len(s) - 1, -1, -1):
            if s[i] == '0':
                ans_length += 1
            else:
                if current_val + power <= k:
                    current_val += power
                    ans_length += 1
                
            if power <= k: 
                power *= 2
            elif power > k and s[i] == '1' and i > 0: 
                # This condition is crucial: if power is already too large for 'k', 
                # and we encounter a '1', it means we cannot add any more '1's.
                # However, we still need to add any remaining '0's to the length.
                # The 'power' variable should not grow further unnecessarily.
                # But it should grow if it's still possible to reach 'k'.
                # The 'if power <= k' above already handles this.
                # The additional check `and s[i] == '1' and i > 0` is redundant
                # if power just gets multiplied always.
                # Let's simplify and just let power always multiply. Python handles large ints.
                pass
        
        return ans_length